appliance/mariadb/process.go

```diff
@@ -0,0 +1,870 @@
+	if err := p.writeConfig(configData{}); err != nil {
+		return "", err
+	}
+	if err := p.unpackXbstream(r); err != nil {
+		return "", err
+	}
+	gtid, err := p.extractBackupGTID()
+	if err != nil {
+		return "", err
+	}
+	if err := p.restoreApplyLog(); err != nil {
+		return "", err
+	}
+	return gtid, nil
+}
+
+func (p *Process) unpackXbstream(r io.ReadCloser) error {
+	defer r.Close()
+
+	cmd := exec.Command(filepath.Join(p.BinDir, "xbstream"), "-x", "--directory="+p.DataDir)
+	cmd.Stdin = r
+
+	if buf, err := cmd.CombinedOutput(); err != nil {
+		p.Logger.Error("xbstream failed", "err", err, "output", string(buf))
+		return err
+	}
+	return nil
+}
+
+func (p *Process) restoreApplyLog() error {
+	cmd := exec.Command(
+		filepath.Join(p.BinDir, "innobackupex"),
+		"--defaults-file="+p.ConfigPath(),
+		"--apply-log",
+		p.DataDir,
+	)
+	if buf, err := cmd.CombinedOutput(); err != nil {
+		p.Logger.Error("innobackupex apply-log failed", "err", err, "output", string(buf))
+		return err
+	}
+	return nil
+}
+
+func (p *Process) assumeStandby(upstream, downstream *discoverd.Instance) error {
+	logger := p.Logger.New("fn", "assumeStandby", "upstream", upstream.Addr)
+	logger.Info("starting up as standby")
+
+	if err := p.writeConfig(configData{ReadOnly: true}); err != nil {
+		logger.Error("error writing config", "path", p.ConfigPath(), "err", err)
+		return err
+	}
+
+	var slaveXLog string
+	if p.running() {
+		if err := p.stop(); err != nil {
+			return err
+		}
+	} else {
+		if p.WaitUpstream {
+			if err := p.waitForUpstream(upstream); err != nil {
+				return err
+			}
+		}
+
+		if err := func() error {
+			logger.Info("retrieving backup")
+			resp, err := http.Get(fmt.Sprintf("http://%s/backup", upstream.Meta["HTTP_ADDR"]))
+			if err != nil {
+				logger.Error("error connecting to upstream for backup", "err", err)
+				return err
+			} else if resp.StatusCode != http.StatusOK {
+				resp.Body.Close()
+				logger.Error("error code returned from backup", "status_code", resp.StatusCode)
+				return err
+			}
+
+			logger.Info("restoring backup")
+			slaveXLog, err = p.Restore(resp.Body)
+			if err != nil {
+				logger.Error("error restoring backup", "err", err)
+				return err
+			}
+			return nil
+		}(); err != nil {
+			if files, err := ioutil.ReadDir("/data"); err == nil {
+				for _, file := range files {
+					os.RemoveAll(filepath.Join("/data", file.Name()))
+				}
+			}
+			return err
+		}
+	}
+
+	if err := p.start(); err != nil {
+		return err
+	}
+
+	if err := func() error {
+		// Connect to local server and set up slave replication.
+		db, err := p.connectLocal()
+		if err != nil {
+			logger.Error("error acquiring connection", "err", err)
+			return err
+		}
+		defer db.Close()
+
+		host, port, _ := net.SplitHostPort(upstream.Addr)
+		logger.Info("changing master", "host", host, "port", port)
+		if _, err := db.Exec(fmt.Sprintf(`SET GLOBAL gtid_slave_pos = "%s";`, slaveXLog)); err != nil {
```

Comment:
Turns out this is necessary, else the slave starts trying to replicate from the start of the log.. which isn't so great for obvious reasons.